src/execution_plan/ops/op_apply_multiplexer.c

```diff
@@ -0,0 +1,150 @@
+OpResult OpApplyMultiplexerReset(OpBase *opBase);
+Record OpApplyMultiplexerConsume(OpBase *opBase);
+void OpApplyMultiplexerFree(OpBase *opBase);
+
+static Record _pullFromBranchStream(OpApplyMultiplexer *apply_multiplexer, int branch_index) {
+	// Propegate record to the top of the match stream.
+	Argument_AddRecord(apply_multiplexer->branches_arguments[branch_index - 1],
+					   Record_Clone(apply_multiplexer->r));
+	return OpBase_Consume(apply_multiplexer->op.children[branch_index]);
+}
+
+static Record OpApplyMultiplexer_OrConsume(OpBase *opBase) {
+	OpApplyMultiplexer *op = (OpApplyMultiplexer *)opBase;
+	while(true) {
+		// Try to get a record from bound stream.
+		op->r = OpBase_Consume(op->bound_branch);
+		if(!op->r) return NULL; // Depleted.
+
+		// Try to get a record from some stream.
+		for(int i = 1; i < op->op.childCount; i++) {
+			Record branch_record = _pullFromBranchStream(op, i);
+			if(branch_record) {
+				// Don't care about the branch record.
+				Record_Free(branch_record);
+				Record r = op->r;
+				op->r = NULL;   // Null to avoid double free.
+				return r;
+			}
+		}
+		// Did not managed to get a record from any branch, loop back and restart.
+		Record_Free(op->r);
+		op->r = NULL;
+	}
+}
+
+static Record OpApplyMultiplexer_AndConsume(OpBase *opBase) {
+	OpApplyMultiplexer *op = (OpApplyMultiplexer *)opBase;
+	while(true) {
+		// Try to get a record from bound stream.
+		op->r = OpBase_Consume(op->bound_branch);
+		if(!op->r) return NULL; // Depleted.
+
+		// Try to get a record from some stream.
+		for(int i = 1; i < op->op.childCount; i++) {
+			Record branch_record = _pullFromBranchStream(op, i);
+			// Don't care about the branch record.
+			if(branch_record) Record_Free(branch_record);
+			else {
+				// Did not managed to get a record from some branch, loop back and restart.
+				Record_Free(op->r);
+				op->r = NULL;
+				break;
+			}
+		}
+		// All branches returned record => all filters are satisfied by the bounded record.
+		Record r = op->r;
+		op->r = NULL;   // Null to avoid double free.
+		return r;
+	}
+}
+
+OpBase *NewApplyMultiplexerOp(ExecutionPlan *plan, AST_Operator boolean_operator) {
+
+	OpApplyMultiplexer *op = rm_calloc(1, sizeof(OpApplyMultiplexer));
+	op->boolean_operator = boolean_operator;
+	// Set our Op operations
+	if(boolean_operator == OP_OR) {
+		OpBase_Init((OpBase *)op, OPType_APPLY_MULTIPLEXER, "OR Apply Multiplexer", OpApplyMultiplexerInit,
+					OpApplyMultiplexer_OrConsume, OpApplyMultiplexerReset, NULL, OpApplyMultiplexerFree, false, plan);
+	} else if(boolean_operator == OP_AND) {
+		OpBase_Init((OpBase *)op, OPType_APPLY_MULTIPLEXER, "AND Apply Multiplexer", OpApplyMultiplexerInit,
+					OpApplyMultiplexer_AndConsume, OpApplyMultiplexerReset, NULL, OpApplyMultiplexerFree, false, plan);
+	} else {
+		assert(false);
+	}
+	return (OpBase *) op;
+}
+
+// Sorts the multiplexer children. Apply operations to the end, filter operations to the start.
```

Comment:
Comment should be more descriptive, explaining that Filter ops evaluate quickly and thus it's more efficient to have them first.

Reply:
done